engine/runtime/src/main/scala/org/enso/compiler/pass/analyse/AmbiguousImportsAnalysis.scala

```diff
@@ -0,0 +1,430 @@
+    copyOfIr
+
+  /** @inheritdoc
+    */
+  override def runModule(
+    ir: IR.Module,
+    moduleContext: ModuleContext
+  ): IR.Module = {
+    val module = moduleContext.module
+    if (module.isSynthetic) {
+      ir
+    } else {
+      val bindingMap = ir.unsafeGetMetadata(
+        BindingAnalysis,
+        "BindingMap should already be present"
+      )
+      val encounteredSymbols = new EncounteredSymbols()
+      ir.copy(
+        imports = ir.imports.flatMap(imp => {
+          analyseAmbiguousSymbols(
+            imp,
+            module,
+            bindingMap,
+            encounteredSymbols
+          ) match {
+            case Left(errors) =>
+              errors
+            case Right(imp) => List(imp)
+          }
+        })
+      )
+    }
+  }
+
+  /** Analyses ambiguous symbols in the given import.
+    * @param imp current import
+    * @param module current module
+    * @param bindingMap binding map of the current module
+    * @return A list of errors, if any encountered, or an import IR, potentially with some
+    *         warnings attached.
+    */
+  private def analyseAmbiguousSymbols(
+    imp: IR.Module.Scope.Import,
+    module: Module,
+    bindingMap: BindingsMap,
+    encounteredSymbols: EncounteredSymbols
+  ): Either[List[IR.Error.ImportExport], IR.Module.Scope.Import] = {
+    imp match {
+      // Import multiple symbols
+      case moduleImport @ IR.Module.Scope.Import.Module(
+            _,
+            _,
+            _,
+            Some(onlyNames),
+            _,
+            _,
+            false,
+            _,
+            _
+          ) =>
+        getImportTarget(moduleImport, bindingMap) match {
+          case Some(importTarget) =>
+            val encounteredErrors: ListBuffer[IR.Error.ImportExport] =
+              ListBuffer()
+            val imp =
+              onlyNames.foldLeft(moduleImport: IR.Module.Scope.Import) {
+                case (imp, symbol) =>
+                  val symbolName = symbol.name
+                  importTarget.resolveExportedSymbol(symbolName) match {
+                    case Right(resolvedName) =>
+                      val symbolPath = resolvedName.qualifiedName.toString
+                      tryAddEncounteredSymbol(
+                        module,
+                        encounteredSymbols,
+                        imp,
+                        symbolName,
+                        symbolPath
+                      ) match {
+                        case Left(error) =>
+                          encounteredErrors += error
+                          imp
+                        case Right(imp) => imp
+                      }
+                    case Left(resolutionError) =>
+                      throw new CompilerError(
+                        s"Unreachable: (should have been resolved in previous passes) $resolutionError"
+                      )
+                  }
+              }
```

Comment:
This is a slightly weird pattern - `foldLeft` only to always end up with `moduleImport`. Why not just `flatMap` on `onlyNames` instead and avoid creating `encounteredErrors` all together?

Reply:
Note that I implemented it via `flatMap` before and I have just found no other way to collect the warnings.